source/common/stats/stats_impl.cc

```diff
@@ -150,10 +150,20 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
 }
 
 RawStatData* HeapRawStatDataAllocator::alloc(const std::string& name) {
-  // This must be zero-initialized
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  absl::string_view key = name;
   RawStatData* data = static_cast<RawStatData*>(::calloc(RawStatData::size(), 1));
-  data->initialize(name);
-  return data;
+  data->initialize(key);
+  auto ret = stats_.insert(data);
```

Comment:
you can take the lock after the call to initialize(), to minimize the time spent holding the lock.  Actually I think you can also let it go immediately after the call to insert as well as ref_count_ is atomic.